code/headers/hardware_usart.hpp

```diff
@@ -5,151 +5,254 @@
-                HWLIB_PANIC_WITH_LOCATION;
-            }
+        hardware_usart_c(unsigned int baudrate) {
+            // set the peripheral for the usart pins
+            set_peripheral<Bus::rx>();
+            set_peripheral<Bus::tx>();
+
+            // enable the peripheral clock for the usart
+            PMC->PMC_PCER0 = (0x01 << detail::periph_id::id<Bus>);
+
+            // disable all interrupts for the usart
+            detail::usart::port<Bus>->US_IDR = 0xFFFFFFFF;
 
-            const auto &curr = usart[uint8_t(usart_port)];
+            // clear the status register for old interrupts
+            (void)detail::usart::port<Bus>->US_CSR;
 
-            hardware_usart = curr.usart;
-            set_peripheral(curr.pio, curr.rx_mask, curr.periph);
-            set_peripheral(curr.pio, curr.tx_mask, curr.periph);
-            PMC->PMC_PCER0 = (0x01 << curr.id);
+            // use the periph_id since that is the same id as the IRQn_type
+            constexpr auto iqrn =
+                static_cast<IRQn_Type>(detail::periph_id::id<Bus>);
+
+            // enable the nvic interrupt for the current usart
+            NVIC_SetPriority(iqrn, 9);
+            NVIC_EnableIRQ(iqrn);
 
             disable();
 
-            hardware_usart->US_BRGR = (5241600u / baudrate);
+            // write the baud rate in the BRGR register
+            detail::usart::port<Bus>->US_BRGR = (5241600u / baudrate);
 
-            hardware_usart->US_MR =
-                    UART_MR_PAR_NO | UART_MR_CHMODE_NORMAL | US_MR_CHRL_8_BIT;
-            hardware_usart->US_IDR = 0xFFFFFFFF;
+            // set the usart mode
+            detail::usart::port<Bus>->US_MR =
+                UART_MR_PAR_NO | UART_MR_CHMODE_NORMAL | US_MR_CHRL_8_BIT;
 
             enable();
-        }
 
-        /// @brief char output operator
-        ///
-        /// Although calling send_byte should do the exact same thing.
-        /// In practice useing this fuction is more stable
-        /// Especially when the values are repeated.
-        hardware_usart_c &operator<<(uint8_t byte) {
-            send_byte(byte);
-            return *this;
+            // enable the interrupt for handling the receiving of data
+            detail::usart::port<Bus>->US_IER = US_IER_RXRDY;
         }
 
-        /// @brief string output operator
-        ///
-        /// used for char arrays to quickly send more than one byte
-        /// In practice useing this fuction is more stable than repeated use of
-        /// send_byte() Especially when the values are repeated.
-        hardware_usart_c &operator<<(const char *c) {
-            for (const char *p = c; *p != '\0'; p++) {
-                send_byte(*p);
+        /**
+         * @brief Write a uint8_t with the usart
+         *
+         * @param c data to send
+         * @return true if the data is sended correctly
+         * @return false if the data has not been sended correctly
+         */
+        bool send(const uint8_t c) override {
```

Comment:
Documentation doesn't seem correct, you either stay in an endless while loop or return true, never a false.